packages/io/x3d-serializer/index.js

```diff
@@ -33,102 +32,119 @@ const stringify = require('onml/lib/stringify')
+  if (objects.length !== objects3d.length) console.warn('some objects could not be serialized to X3D')
+
+  options.statusCallback && options.statusCallback({ progress: 0 })
 
   // construct the contents of the XML
-  var body = ['X3D',
+  let body = ['X3D',
     {
       profile: 'Interchange',
       version: '3.3',
       'xmlns:xsd': 'http://www.w3.org/2001/XMLSchema-instance',
       'xsd:noNamespaceSchemaLocation': 'http://www.web3d.org/specifications/x3d-3.3.xsd'
     },
     ['head', {},
-      ['meta',{name: 'creator', content: 'Created using JSCAD'}]
+      ['meta', { name: 'creator', content: 'Created using JSCAD' }]
     ]
   ]
-  body = body.concat(convertObjects(objects, options))
+  body = body.concat(convertObjects(objects3d, options))
 
   // convert the contents to X3D (XML) format
-  var contents = `<?xml version="1.0" encoding="UTF-8"?>
+  let contents = `<?xml version="1.0" encoding="UTF-8"?>
 ${stringify(body)}`
 
-  options && options.statusCallback && options.statusCallback({progress: 100})
+  options && options.statusCallback && options.statusCallback({ progress: 100 })
 
   return [contents]
 }
 
 const convertObjects = (objects, options) => {
   let scene = ['Scene', {}]
   let shapes = []
-  objects.forEach(function (object, i) {
-    options.statusCallback && options.statusCallback({progress: 100 * i / objects.length})
-    if (isCSG(object) && object.polygons.length > 0) {
-      object = ensureManifoldness(object)
-      shapes.push(convertCSG(object, options))
+  objects.forEach((object, i) => {
+    options.statusCallback && options.statusCallback({ progress: 100 * i / objects.length })
+
+    if (geometry.geom3.isA(object)) {
+      let polygons = geometry.geom3.toPolygons(object)
+      if (polygons.length > 0) {
+        // TODO object = ensureManifoldness(object)
+        shapes.push(convertShape(object, options))
+      }
     }
   })
   scene = scene.concat(shapes)
   return [scene]
 }
 
-const convertCSG = (object, options) => {
-  var shape = ['Shape', {}, convertMesh(object, options)]
+const convertShape = (object, options) => {
+  let shape = ['Shape', {}, convertMesh(object, options)]
   return shape
 }
 
 const convertMesh = (object, options) => {
-  let mesh = object.toTriangles()
+  let mesh = convertToTriangles(object, options)
   let lists = polygons2coordinates(mesh, options)
 
   let indexList = lists[0].join(' ')
   let pointList = lists[1].join(' ')
   let colorList = lists[2].join(' ')
 
-  var faceset = [
+  let faceset = [
     'IndexedTriangleSet',
-    {ccw: 'true', colorPerVertex: 'false', solid: 'false', index: indexList},
-    ['Coordinate', {point: pointList}],
-    ['Color', {color: colorList}]
+    { ccw: 'true', colorPerVertex: 'false', solid: 'false', index: indexList },
+    ['Coordinate', { point: pointList }],
+    ['Color', { color: colorList }]
   ]
   return faceset
 }
 
+const convertToTriangles = (object, options) => {
```

Comment:
nice !
But I am a bit worried about this one: do we not have a more generic need for triangulated meshes ? (that is needed from webgl viewing , most 3d exports etc), so put another way: should we not have a geom3.toTriangles ? (that can use this code for example ?)

Reply:
I think that a toTriangles() function would be a better solution. As we move to a 'descriptive' model then direct access to the polygons should not be possible. Conversion of the 'description' model to data should be requested via a function call, i.e. toPolygons(), toTriangles(), etc.

Not sure if we should do this now. SIGH

Reply:
Agreed but It is not urgent, I will add it to my "V2 changes and tweaks" doc for now :-)